finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+    authentications: Vec<Authentication>,
+    peer_id: PeerId,
+) -> (DiscoveryMessage, DataCommand) {
+    (
+        DiscoveryMessage::Authentications(authentications),
+        DataCommand::SendTo(peer_id, Protocol::Generic),
+    )
+}
+
+impl Discovery {
+    /// Create a new discovery handler with the given response/broadcast cooldown.
+    pub fn new(cooldown: time::Duration) -> Self {
+        Discovery {
+            cooldown,
+            last_broadcast: HashMap::new(),
+            last_response: HashMap::new(),
+            requested_authentications: HashMap::new(),
+            next_query: NodeIndex(0),
+        }
+    }
+
+    /// Returns messages that should be sent as part of authority discovery at this moment.
+    pub fn discover_authorities(
+        &mut self,
+        handler: &SessionHandler,
+    ) -> Vec<(DiscoveryMessage, DataCommand)> {
+        let missing_authorities = handler.missing_nodes();
+        if missing_authorities.is_empty() {
+            return Vec::new();
+        }
+        let node_count = handler.node_count();
+        let authentication = handler.authentication();
+        if missing_authorities.len() * 3 > 2 * node_count.0 {
+            // We know of fewer than 1/3 authorities, broadcast our authentication and hope others
+            // respond in kind.
+            vec![authentication_broadcast(authentication)]
+        } else {
+            // Attempt learning about more authorities from the ones you already know.
+            let mut result = Vec::new();
+            let mut target = self.next_query;
+            while result.len() < 2 {
+                if let Some(peer_id) = handler.peer_id(&target) {
+                    result.push(request(
+                        missing_authorities.clone(),
+                        authentication.clone(),
+                        peer_id,
+                    ));
+                }
+                target = NodeIndex((target.0 + 1) % node_count.0);
+            }
+            self.next_query = target;
+            result
+        }
+    }
+
+    fn handle_authentication(
+        &mut self,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> Vec<Multiaddr> {
+        if !handler.handle_authentication(authentication.clone()) {
+            return Vec::new();
+        }
+        authentication.0.address()
+    }
+
+    fn handle_broadcast(
+        &mut self,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> (Vec<Multiaddr>, Option<(DiscoveryMessage, DataCommand)>) {
+        let addresses = self.handle_authentication(authentication.clone(), handler);
+        if addresses.is_empty() {
+            return (addresses, None);
+        }
+        let node_id = authentication.0.creator();
+        let rebroadcast = match self.last_broadcast.get(&node_id) {
+            Some(instant) => time::Instant::now() > *instant + self.cooldown,
+            None => true,
+        };
```

Comment:
- do I understand correctly: if the last broadcast from `node_id` was long time ago, we are trying to help them?
 - what about a tiny pretty auxiliary method `should_rebroadcast()`?

Reply:
If it was long ago then we can do it again, so we do. It's just a spam/congestion avoidance scheme.

Added auxiliary method.